### ISSUE TITLE
Page/Form/Style transforms

### ISSUE THREAD

**Issue by Bloke** [textpattern/textpattern#915]
Now that we've moved the Textfilter options inline on the Write panel, and filesystem support of Themes is cooking, I'd like to revisit the notion of transforms in other areas of the interface, e.g. Page/Forms/Stylesheets.

Essentially, this is something that [net-carver and I discussed](https://forum.textpattern.io/viewtopic.php?pid=251000#p251000) years ago, but the codebase wasn't mature enough to support it back then. The theory is this:

* Introduce in the txp_page/txp_form/txp_css tables two extra columns: 'raw' (for holding content) and 'transform' for holding the id of a transformation that's applied to it.
* Use that column as content in the textareas when managing the UI or during export of Theme assets.
* Add an entirely optional callback to register Textfilters for each of these panels too, so plugins can offer "transforms" that can be applied to the content at Save time.
* If there are any transforms defined, the list of available transforms appear above the textarea as a dropdown.
* Upon Save the 'raw' content is passed through the chosen Textfilter and the output stored in the (current) data column in the table. Thus it becomes a "cooked" variant of the data in the "raw" field. The chosen transform is stored alongside the data so it can be used next time the asset is loaded.
* Upon import, if the textfilter is defined somehow (maybe even as part of the filename?) then it's used to process the incoming data prior to being stored.
* Any time there are no filters in place, the system acts like "leave text untouched" and passes it straight through.

It's a bit of a woolly definition right now, and I'm sure there are gotchas, but such a system appears to have some real-world benefits. Maybe a SASS compiler so you can write CSS in SASS and have it compiled upon Save. Or a minifier. Or both! (I don't know if SASS has a one-to-one relationship with source to compiled output so maybe this isn't realistic if you rely on a series of files to generate your CSS).

Maybe you want to run your Page / Form code through a Txp linter to catch any stray bugs before it's committed to your site? You could do that with a custom Textfilter.

Of course, you could Textile content, apply Markdown to it, use Haml/Yaml/XML to write your templates and have them encoded to HTML on save. Heck, you could even write a filter to inject variables into your templates, replacing placeholder content. Lots of applications.

I'd love to have the ability to chain filters, but I'm not sure if that's easily achievable in the UI.

Anyway, overall is this idea worth pursuing? Not for 4.7.0, some future version.

**Comment by philwareham** [comment]
Sounds interesting.

Not sure how things like Haml would interact with Textpattern tags (I've not used it personally) but probably can be made to work. I prefer to keep all CSS (including Sass) out of the database but again I can see the value to some. Much of this can be achieved in via build tasks within Grunt/Gulp easier in my opinion, so flat file stuff that we've already discussed provides a more immediate solution.

I'd imagine the 'cooked' files could be used in some core cacheing solution in future too?

**Comment by Bloke** [comment]
Yes, if you're comfortable with a filesystem approach, then task automation tools like Gulp are perfect. For the rest of us who haven't caught up and/or prefer to let plugin authors do the coding, the hooks allow a great degree of freedom over what "transform" means. It could be one thing, or a chain of things. Even business logic (update this, decrement that, do something else, grab the results, inject in template). Can't think of an example off the top of my head though.

**Comment by bloatware** [comment]
I like it very much! But no clear idea of implementation atm: should the cooked code be stored in the same or a separate table? should any field be potentially transformable? We could start playing with custom fields in 4.8...

**Comment by Bloke** [comment]
Well we store the cooked stuff in the same table in the `textpattern` table. Not saying that's the right or wrong way to do it. Interesting you mention custom fields, because they really are just going to be an uber-efficient way of storing 'stuff' connected to other tables. Cooked content could indeed be such 'stuff'. In fact, iirc, I did put a transform column in there (because it's primarily conceived for storing Textiled CF content) so it could be an ideal candidate. In which case, we ge this functionality almost for free.
